gstreamer/src/date_time_serde.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2018 François Laignel <fengalin@free.fr>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use serde::de::{Deserialize, Deserializer};
+use serde::ser::{Serialize, SerializeStruct, Serializer};
+use DateTime;
+
+impl<'a> Serialize for DateTime {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut datetime = serializer.serialize_struct("DateTime", 8)?;
+        datetime.serialize_field("tz_offset", &self.get_time_zone_offset())?;
+        datetime.serialize_field("y", &self.get_year())?;
+        datetime.serialize_field("m", &self.get_month())?;
+        datetime.serialize_field("d", &self.get_day())?;
+        datetime.serialize_field("h", &self.get_hour())?;
+        datetime.serialize_field("mn", &self.get_minute())?;
+        datetime.serialize_field("s", &self.get_second())?;
+        datetime.serialize_field("us", &self.get_microsecond())?;
```

Comment:
`DateTime` can have no ms, s, m, etc. Calling the getters if they don't have the field will cause an assertion, and there are the `has_XXX()` functions for checking their existence

Reply:
Oops, I will fix this.

Reply:
This raises additional questions. I'm heading toward this representation:

```
struct DateTimeSerde {
    y: Option<i32>,
    m: Option<i32>,
    time: Option<(i32, i32)>,
    s: Option<f64>,
    tz_offset: Option<f32>,
}
```

- I merged `h` and `mn` in `time` since `has_time` refers to both.
- I used an `Option` for `tz_offset` and I'm planning on using `new_local_time` in case it is `None`.
- They are no constructors for cases where `time` is defined, but `y`, `h` and `m` aren't, so I guess this would be the actual model to be used (schematically):
```
date_time: (
    y: Option<(
        i32,
        m: Option<(
            i32,
            time: Option<(
                i32,
                i32,
                s: Option<f64>,
                tz_offset: Option<f32>,
            )>,
        >,
    )>,
)
```

... but I don't really like it because I don't find it easy to read.

Reply:
Yes, an `enum` seems better again. However, we might want to use abbreviations otherwise we get things like this:

```
YearMonthDayHourMinutesSecondsTzOffset(2018, 5, 28, 16, 6, 42841000, 2)
```

what do you think about this instead:

```
YMDHMnSTz(2018, 5, 28, 16, 6, 42841000, 2)
```

Reply:
Sounds good to me